model/perf.go

```diff
@@ -637,3 +655,174 @@ func (r *PerformanceResults) FindOutdatedRollups(ctx context.Context, name strin
+	Data         []TimeSeriesEntry `bson:"times_series"`
+}
+
+func makeTimeSeriesPipeline(timeSeriesId TimeSeriesId) []bson.M {
+	return []bson.M{
+		{
+			"$match": bson.M{
+				"info.project":   timeSeriesId.Project,
+				"info.variant":   timeSeriesId.Variant,
+				"info.task_name": timeSeriesId.Task,
+				"info.test_name": timeSeriesId.Test,
+			},
+		},
+		{
+			"$match": bson.M{
+				"info.order": bson.M{
+					"$exists": true,
+				},
+				"info.mainline": true,
+			},
+		},
+		{
+			"$unwind": "$rollups.stats",
+		},
+		{
+			"$match": bson.M{
+				"rollups.stats.name": timeSeriesId.Measurement,
+			},
+		},
+		{
+			"$project": bson.M{
+				"value": "$rollups.stats.val",
+				"order": "$info.order",
+			},
+		},
+	}
+}
+
+func GetTimeSeries(ctx context.Context, env cedar.Environment, timeSeriesId TimeSeriesId) (*TimeSeries, error) {
+	timeSeriesAggregator := makeTimeSeriesPipeline(timeSeriesId)
+	cur, err := env.GetDB().Collection(perfResultCollection).Aggregate(ctx, timeSeriesAggregator)
+	if err != nil {
+		return nil, errors.Wrap(err, "Cannot aggregate time series")
+	}
+	defer cur.Close(ctx)
+	var res []TimeSeriesEntry
+	err = cur.All(ctx, &res)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not decode time series")
+	}
+	return &TimeSeries{
+		TimeSeriesId: timeSeriesId,
+		Data:         res,
+	}, nil
+}
+
+func ReplaceChangePoints(ctx context.Context, env cedar.Environment, timeSeries *TimeSeries, changePoints []ChangePoint) []error {
+	errs := []error{}
+	err := ClearChangePoints(ctx, env, timeSeries.TimeSeriesId)
+	if err != nil {
+		errs = append(errs, errors.Wrapf(err, "Unable to clear change points for measurement %s", timeSeries.TimeSeriesId))
+		return errs
+	}
+	for _, cp := range changePoints {
+		perfResultId := timeSeries.Data[cp.Index].PerfResultID
+		err = CreateChangePoint(ctx, env, perfResultId, timeSeries.TimeSeriesId.Measurement, cp.Algorithm)
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "Failed to update performance result with change point %s", perfResultId))
+		}
+	}
+	return errs
+}
+
+func ClearChangePoints(ctx context.Context, env cedar.Environment, timeSeriesId TimeSeriesId) error {
+	seriesFilter := bson.M{
+		"info.project":              timeSeriesId.Project,
+		"info.variant":              timeSeriesId.Variant,
+		"info.task_name":            timeSeriesId.Task,
+		"info.test_name":            timeSeriesId.Test,
+		"rollups.stats.name":        timeSeriesId.Measurement,
+		"change_points.measurement": timeSeriesId.Measurement,
+	}
+	pullUpdate := bson.M{
+		"$pull": bson.M{
+			"change_points": bson.M{
+				"measurement": timeSeriesId.Measurement,
+			},
+		},
+	}
+	_, err := env.GetDB().Collection(perfResultCollection).UpdateMany(ctx, seriesFilter, pullUpdate)
+	return errors.Wrap(err, "Unable to clear change points")
+}
+
+func CreateChangePoint(ctx context.Context, env cedar.Environment, resultToUpdate string, measurement string, algorithm AlgorithmInfo) error {
```

Comment:
didn't we talk about about merging this into the previous function?

Reply:
Clear change points clears them all for a metric, create adds one to a specific point in its history. I added `ReplaceChangePoints` to do both in an idempotent fashion. I'm going to make these private so nobody calls them accidentally from outside of the package.